test/e2e/operator_test.go

```diff
@@ -3579,6 +3579,36 @@ func waitForDeploymentEnvVar(t *testing.T, cl client.Client, deployment *appsv1.
 	return err
 }
 
+// waitForDeploymentCompleteWithCleanup waits for a deployment to complete its rollout, then waits for the old
+// generation's pods to finish terminating.
+func waitForDeploymentCompleteWithCleanup(t *testing.T, cl client.Client, deploymentName types.NamespacedName, timeout time.Duration) error {
+	t.Helper()
+	deployment := &appsv1.Deployment{}
+	if err := cl.Get(context.TODO(), deploymentName, deployment); err != nil {
+		return fmt.Errorf("failed to get deployment %s: %w", deploymentName.Name, err)
+	}
+
+	if deployment.Generation != deployment.Status.ObservedGeneration {
+		if err := waitForDeploymentComplete(t, cl, deployment, timeout); err != nil {
+			return fmt.Errorf("timed out waiting for deployment %s to complete rollout: %w", deploymentName.Name, err)
+		}
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return fmt.Errorf("deployment %s has invalid spec.selector: %w", deploymentName.Name, err)
+	}
+
+	return wait.PollImmediate(2*time.Second, timeout, func() (bool, error) {
+		pods := &corev1.PodList{}
+		if err := cl.List(context.TODO(), pods, client.MatchingLabelsSelector{Selector: selector}); err != nil {
+			t.Logf("failed to list pods for deployment %q: %v", deploymentName.Name, err)
+			return false, nil
+		}
+		return len(pods.Items) == int(*deployment.Spec.Replicas), nil
```

Comment:
Isn't this the same as doing`spec.replicas == status.replicas`, which is what `waitForDeploymentComplete` already does? You are counting the pods manually, but I don't see how this is filtering out terminating pods.

Shouldn't you check for no `metadata.deletionTimestamp` on each of the pods before comparing the pod count to `spec.replicas`?

Reply:
Alternatively, a better solution might be instead to ensure there are no terminating pods that are also ready. This saves about 15 seconds of blocking when there is 1 terminating router that is not ready (it's not going to get requests), and we have our two other router pods up.

Reply:
I think the reason we're seeing this issue is that `status.replicas` seems to either be pruning pods that are terminating or it's only including pods in the current generation. Either way, I've found that `spec.replicas == status.replicas` doesn't say if there are still terminating pods, while manually listing the pods does.

I do like the idea of checking for pods that are terminating but still ready. I'll try that and if it works, I'll add it to the PR.

Reply:
It turns out that "terminating" isn't actually a pod state; oc/kubectl will show that, but it's not a part of the pod status.

However, I did update the function to only count the number of ready pods, so it should return once the old pods become not ready.